### PR TITLE
ENH: Markups plane node for imager and image window

### PR DESCRIPTION
**Pull request by MichaelColonel**
DRR important updates:

1. First commit updates visualization of DRR imager and image sub-window from closed curve to plane node, because it's much better to understand.
2. Second commit makes observation of reference beam node. DRR geometry and transformation parameters such as View-Up and normal vectors are updated if reference beam node transformation or geometry was updated.
3. Use CT data from SampleData module in DRR python test.

### CODE REVIEW

**Comment by MichaelColonel** [comment]
Some updates for DRR computation.

File: DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx [hunk 1]
```diff
@@ -289,8 +299,10 @@ void qSlicerDrrImageComputationPlastimatchParametersWidget::onUseExponentialMapp
     qCritical() << Q_FUNC_INFO << ": Invalid parameter node";
     return;
   }
-
+  d->ParameterNode->DisableModifiedEventOn();
   d->ParameterNode->SetExponentialMappingFlag(value);
+  d->ParameterNode->DisableModifiedEventOn();
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx]: This should be DisableModifiedEventOff. Why is it necessary invoke DisableModifiedEventOn/DisableModifiedEventOff for all of these functions?

Comment (MichaelColonel) [reply DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx]: Old version. Obsolete code. I have removed all unnecessary node update blocks, except one which is not part of Plastimatch DRR program.

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx [hunk 2]
```diff
@@ -700,15 +620,30 @@ void vtkSlicerDrrImageComputationLogic::ShowMarkupsNodes(bool toggled)
...
+  {
+    vtkErrorMacro("CreateImagerBoundary: Invalid MRML scene");
+    return nullptr;
+  }
+  vtkMRMLNode* node = scene->AddNewNodeByClass( "vtkMRMLMarkupsPlaneNode", IMAGER_BOUNDARY_MARKUPS_NODE_NAME);
+
+  vtkMRMLMarkupsPlaneNode* imagerMarkupsNode = nullptr;
+  if (node)
+  {
+    imagerMarkupsNode = vtkMRMLMarkupsPlaneNode::SafeDownCast(node);
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: The node doesn't need to be checked twice you can call SafeDownCast and AddNewNodeByClass in the same line.*

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx [hunk 3]
```diff
@@ -754,15 +688,30 @@ vtkMRMLMarkupsClosedCurveNode* vtkSlicerDrrImageComputationLogic::CreateImagerBo
...
-  this->GetMRMLScene()->AddNode(imageWindowMarkupsNode);
-  imageWindowMarkupsNode->SetName(IMAGE_WINDOW_MARKUPS_NODE_NAME);
-  imageWindowMarkupsNode->SetCurveTypeToLinear();
+  vtkMRMLScene* scene = this->GetMRMLScene(); 
+  if (!scene)
+  {
+    vtkErrorMacro("CreateImageWindow: Invalid MRML scene");
+    return nullptr;
+  }
+  vtkMRMLNode* node = scene->AddNewNodeByClass( "vtkMRMLMarkupsPlaneNode", IMAGE_WINDOW_MARKUPS_NODE_NAME);
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: *Same here.

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx [hunk 4]
```diff
@@ -821,12 +769,28 @@ vtkMRMLMarkupsClosedCurveNode* vtkSlicerDrrImageComputationLogic::CreateImageWin
...
-  vtkNew<vtkMRMLMarkupsLineNode> vectorMarkupsNode;
-  this->GetMRMLScene()->AddNode(vectorMarkupsNode);
-  vectorMarkupsNode->SetName(NORMAL_VECTOR_MARKUPS_NODE_NAME);
+  vtkMRMLScene* scene = this->GetMRMLScene(); 
+  if (!scene)
+  {
+    vtkErrorMacro("CreateImagerNormal: Invalid MRML scene");
+    return nullptr;
+  }
+  vtkMRMLNode* node = scene->AddNewNodeByClass( "vtkMRMLMarkupsLineNode", NORMAL_VECTOR_MARKUPS_NODE_NAME);
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: *Same here.

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx [hunk 5]
```diff
@@ -855,12 +819,28 @@ vtkMRMLMarkupsLineNode* vtkSlicerDrrImageComputationLogic::CreateImagerNormal(vt
...
-  vtkNew<vtkMRMLMarkupsLineNode> vectorMarkupsNode;
-  this->GetMRMLScene()->AddNode(vectorMarkupsNode);
-  vectorMarkupsNode->SetName(VUP_VECTOR_MARKUPS_NODE_NAME);
+  vtkMRMLScene* scene = this->GetMRMLScene(); 
+  if (!scene)
+  {
+    vtkErrorMacro("CreateImagerVUP: Invalid MRML scene");
+    return nullptr;
+  }
+  vtkMRMLNode* node = scene->AddNewNodeByClass( "vtkMRMLMarkupsLineNode", VUP_VECTOR_MARKUPS_NODE_NAME);
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: *Same here.

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx [hunk 6]
```diff
@@ -901,12 +881,28 @@ vtkMRMLMarkupsLineNode* vtkSlicerDrrImageComputationLogic::CreateImagerVUP(vtkMR
...
-  vtkNew<vtkMRMLMarkupsFiducialNode> pointsMarkupsNode;
-  this->GetMRMLScene()->AddNode(pointsMarkupsNode);
-  pointsMarkupsNode->SetName(FIDUCIALS_MARKUPS_NODE_NAME);
+  vtkMRMLScene* scene = this->GetMRMLScene(); 
+  if (!scene)
+  {
+    vtkErrorMacro("CreateFiducials: Invalid MRML scene");
+    return nullptr;
+  }
+  vtkMRMLNode* node = scene->AddNewNodeByClass( "vtkMRMLMarkupsFiducialNode", FIDUCIALS_MARKUPS_NODE_NAME);
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: *Same here.

File: DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx [hunk 7]
```diff
@@ -317,8 +315,9 @@ void qSlicerDrrImageComputationPlastimatchParametersWidget::onHUThresholdChanged
     qCritical() << Q_FUNC_INFO << ": Invalid parameter node";
     return;
   }
-
+  d->ParameterNode->DisableModifiedEventOn();
```
Comment (Sunderlandkyl) [inline DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx:318]: Is this one necessary?

**Review by Sunderlandkyl — COMMENTED** [review]
Thanks, I've added a few comments.

Couple of questions:
- What is the behavior when loading a scene that uses the old curve nodes?
- Are there any new tests failing?

File: DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx (hunk 7 above)
Comment (MichaelColonel) [reply DrrImageComputation/Widgets/qSlicerDrrImageComputationPlastimatchParametersWidget.cxx:318]: It's recommended. Every time when the HU threshold value is changed, the vtkCommand::ModifiedEvent is invoked and it updates markups geometry, which it quite slow. In that case GUI update becomes sluggish.

File: DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx (hunk 2 above)
Comment (MichaelColonel) [reply DrrImageComputation/Logic/vtkSlicerDrrImageComputationLogic.cxx]: I will fix that.

**Comment by MichaelColonel** [comment]
> * Are there any new tests failing?

The http://slicer.kitware.com/midas3 site is no longer available. Tests can't download the data anymore, that is why some python tests are failing recently. DRR test uses the CTChest sample data anyway, now it's just a build-in data in the 3D Slicer.

**Comment by MichaelColonel** [comment]
> * What is the behavior when loading a scene that uses the old curve nodes?

Major difference is a visual representation, everything else is the same.
Before is a closed curve markups. After is a new plane markups.
Plane markups is better (IMHO).
image

**Comment by Sunderlandkyl** [comment]
Thanks, it looks good to me.

**Pull request merged**